catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/resource/download/ReliableResourceDownloadManager.java

```diff
@@ -200,21 +210,100 @@ public void setProductCacheDirectory(String productCacheDirectory) {
+        //if the cache is not pending for this key, then create a new reliable resource using that key.
+        //Also, tell the resource cache through the downloader config that we have a new pending item.
+        if (!downloaderConfig.getResourceCache()
+                .isPending(key)) {
+
+            // Fully qualified path to cache file that will be written to.
+            // Example:
+            // <INSTALL-DIR>/data/product-cache/<source-id>-<metacard-id>
+            // <INSTALL-DIR>/data/product-cache/ddf.distribution-abc123
+            filePath = Paths.get(downloaderConfig.getResourceCache()
+                    .getProductCacheDirectory(), key);
+            managerReliableResource = new ReliableResource(key,
+                    filePath.toString(),
+                    resource.getMimeType(),
+                    resource.getName(),
+                    metacard);
+
+            try {
+                cacheFile = filePath.toFile();
+                if (cacheFile.canWrite() || cacheFile.createNewFile()) {
+                    downloaderConfig.getResourceCache()
+                            .addPendingCacheEntry(managerReliableResource);
+                    doCaching = true;
+                }
+            } catch (IOException e) {
+                LOGGER.error("Unable to open cache file {} - no caching will be done.", filePath);
+                cacheFile = null;
+            }
+
+        } else {
+            LOGGER.debug("Cache key {} is pending caching", key);
+        }
+
+        boolean continueDownloadingWhenCancelled =
+                doCaching && downloaderConfig.isCacheWhenCanceled();
+
+        resourceResponse = downloader.setupDownload(metacard,
+                cacheFile,
+                continueDownloadingWhenCancelled);
 
+        downloadStatusInfo.addDownloadInfo(downloadIdentifier, downloader, resourceResponse);
+
+        // Start download in separate thread so can return ResourceResponse with
+        // ReliableResourceInputStream available for client to start reading from
+        // downloaderConfig.getExecutor().submit(downloader);
         // Start download in separate thread so can return ResourceResponse with
         // ReliableResourceInputStream available for client to start reading from
-        executor.submit(downloader);
+
+        ListenableFuture downloadFuture = executor.submit(downloader);
+
+        //if there is additional caching activity that is needed, such as determining the pending status, the
+        // ResourceDownloadCallback is used
+        if (doCaching) {
+            Futures.addCallback(downloadFuture,
+                    new ResourceDownloadCallback(downloader,
+                            cacheFile,
+                            managerReliableResource,
+                            downloaderConfig.getResourceCache()));
+        }
 
         // Wait for download to get started before returning control to client
         Stopwatch stopwatch = Stopwatch.createStarted();
```

Comment:
Can we try to replace the `downloadStarted` `AtomicBoolean` with a `CountdownLatch`? It would allow us to get rid of this busy and the code would look _much_ nicer.